### DISCUSSION TITLE
GROOPS- Station velocities calculation

### DISCUSSION THREAD

**Discussion by Bemtgenj** [groops-devs/groops#100] [Q&A] [answered]
Hello,

Does GROOPS provide a script that can calculate GNSS station velocities in a time series (similar to the MIDAS software from the University of Nevada, Reno. I have looked through the documentation and have not found anything like it (maybe i missed it or did the wrong search)?
Thank you for your help

Cheers
Jean

**Comment by tmayerguerr** [comment, marked as answer]
Hi Jean,

There is no direct program for this. Unfortunately, I don't know what the MIDAS software can do.

A file with the position time series can be generated from a list of files with [InstrumentConcatenate](https://groops-devs.github.io/groops/html/InstrumentConcatenate.html).

Epoch-wise velocities could be calculated with [InstrumentFilter](https://groops-devs.github.io/groops/html/InstrumentFilter.html) (with the Derivative filter). For noisy data, I would use a polynomial of degree 1 and then apply an additional low-pass filter.

Alternatively, piecewise constant velocities could be calculated with [InstrumentArcCalculate](https://groops-devs.github.io/groops/html/InstrumentArcCalculate.html) using leastSquare (setup parameters in each line px,vx,py,vy,pz,vz)
and three leastSquares lines:
- data0 - (px+vx*(epoch-51544))
- data1 - (py+vy*(epoch-51544))
- data2 - (pz+vz*(epoch-51544))
with the arbitrary initial date mjd=51544 (2000-01-01)

The outColumns are the estimated velocities: vx, vy,vz.

In both cases, the time series must be split into arcs with [InstrumentSynchronize](https://groops-devs.github.io/groops/html/InstrumentSynchronize.html) beforehand.
For [InstrumentFilter](https://groops-devs.github.io/groops/html/InstrumentFilter.html) there must be no data gaps in the arcs (minGap empty). A time series with discontinuities can be read in with timeIntervals->instrument (However you generate it). Both methods needs a minimum of epochs within an arc (period between gaps). Remove short periods with minArcLength.

I hope this gives an idea how to calculate velocities.

regards
Torsten

(Edited to fix the errors mentioned below)

**Reply by Bemtgenj** [reply]
Hi Torsten,

Thank you very much for the  quick feedback. Correct - this gives me an idea on how to start / do the calculations. Let me try it out and if I have more questions I will get back to you.

Cheers 
Jean

**Comment by Bemtgenj** [comment]
Hi Torsten

short update on trying out the different solutions

- InstrumentFilter - I get an error that the the timeseries is too short (<1> elements) to apply a filter with a warmup length of <4>. For my tryout is used zim2 station 2020/01/01->2020/02/29 (i also got the same message when i did 2020/01/01->2020/04/30)  would i need a period of about 1 year to get enough points for the calculation?
- InstrumentArcCalculate-i get the following message in the least square: in C:\groops\source\parser\expressionParser.cpp:1503 (derivative): Expression(leastSquares = 'data0 - (px + vx*(epoch - date2mjd(2000, 1, 1)))')
Derivative not defined for "data2mjd". or when i do replace the date2mjd (2000,1,1) with 51544 i get a message: in C:\groops\source\base\matrix.cpp:1197 (leastSquares): min[ A(3 x 6)*x - l(3 x 1) Dimension error

I am not yet sure i completely understand the least square line equation yet.

Txs for the help
Cheers
Jean

**Reply by tmayerguerr** [reply]
Both methods needs a minimum of epochs within an arc (period between gaps). You can remove short periods (short arcs) in [InstrumentSynchronize](https://groops-devs.github.io/groops/html/InstrumentSynchronize.html) with minArcLength.

Ok, I missed that  date2mjd works not within leastSquares.

LeastSquares performs a least squares fit to minimize the given expression and estimates the unknown parameters. In this case the equation 'data0 - (px + vx*(epoch - 51544))' fits a regression line to data0 and estimates the start position px at mjd=51544 and the grade (velocity) vx [m/day]. Same for x,y,z coordinates. This adjustment is performed for each arc. To estimate a straight line you need at least to epochs within an arc.

regards
Torsten

**Comment by Bemtgenj** [comment]
Hi Torsten

Txs for the feedback - i will try that

Cheers
Jean

**Comment by Bemtgenj** [comment]
Hi Torsten

the script worked fine. I do have one small question - why is the date in the date in the least square that you mentioned 2000-01-01; is that just as an example or was a specific reason for that? Could i use my initial date for the time series instead?

Thank you for your continuing help - very much appreciated

Cheers
Jean

**Reply by tmayerguerr** [reply]
You are right, the date is just an example. It defines the reference epoch of the initial position (px,py,pz). You can use your initial date instead.

**Comment by Bemtgenj** [comment]
Hi Torsten

txs for the clarification.

Cheers
Jean

**Reply by hadihaseri** [reply]
Hi Jean, 

Would you be able to share the script with me? 

Thank you for your help

Kind Regards,
Hadi
